ring/batch.go

```diff
@@ -50,32 +50,56 @@ func isHTTPStatus4xx(err error) bool {
-	return DoBatchWithClientError(ctx, op, r, keys, callback, cleanup, isHTTPStatus4xx)
+	return DoBatchWithOptions(ctx, op, r, keys, callback, DoBatchOptions{
+		Cleanup:       cleanup,
+		IsClientError: isHTTPStatus4xx,
+	})
 }
 
-// DoBatchWithClientError request against a set of keys in the ring,
-// handling replication and failures. For example if we want to write
-// N items where they may all hit different instances, and we want them
-// all replicated R ways with quorum writes, we track the relationship
-// between batch RPCs and the items within them.
-//
-// callback() is passed the instance to target, and the indexes of the keys
-// to send to that instance.
-//
-// cleanup() is always called, either on an error before starting the batches
-// or after they all finish.
+// DoBatchOptions defines options for the DoBatchWithOptions call.
+// Zero value options are valid, as well as individual zero valued fields.
+type DoBatchOptions struct {
+	// Cleanup is always called, either on an error before starting the batches or after they are all finished.
+	// If nil, a noop will be called.
+	Cleanup func()
+
+	// IsClientError classifies errors returned by `callback()` into client or server errors.
+	// See `batchTracker.record()` function for details about how errors are combined into final error returned by DoBatchWithClientError.
+	// If nil, a default implementation is used that classifies grpc errors containing status codes 4xx as client errors.
+	IsClientError func(error) bool
+
+	// Go will be used to spawn the callback goroutines, and can be used to use a worker pool like concurrency.ReusableGoroutinesPool.
+	Go func(func())
+}
+
+func (o *DoBatchOptions) setDefaults() {
+	if o.Cleanup == nil {
+		o.Cleanup = func() {}
+	}
+	if o.IsClientError == nil {
+		o.IsClientError = isHTTPStatus4xx
+	}
+	if o.Go == nil {
+		o.Go = func(f func()) { go f() }
+	}
+}
+
+// DoBatchWithOptions request against a set of keys in the ring, handling replication and failures.
+// For example if we want to write N items where they may all hit different instances,
+// and we want them all replicated R ways with quorum writes,
+// we track the relationship between batch RPCs and the items within them.
 //
-// isClientError() classifies errors returned by `callback()` into client or
-// server errors. See `batchTracker.record()` function for details about how
-// errors are combined into final error returned by DoBatchWithClientError.
+// See comments on DoBatchOptions for available options for this call.
 //
 // Not implemented as a method on Ring, so we can test separately.
-func DoBatchWithClientError(ctx context.Context, op Operation, r ReadRing, keys []uint32, callback func(InstanceDesc, []int) error, cleanup func(), isClientError func(error) bool) error {
+func DoBatchWithOptions(ctx context.Context, op Operation, r ReadRing, keys []uint32, callback func(InstanceDesc, []int) error, o DoBatchOptions) error {
```

Comment:
Did you remove `DoBatchWithClientError()`? 
I would put it back and possibly mark it as deprecated, since it will now be part of `r264`.

Reply:
We don't do server in dskit. We're already very generous at leaving the `DoBatch` with the old signature: I didn't change it just  to make the update path easier for codebases with tangled dependencies.

I don't see any need to leave `DoBatchWithClientError` here.

Reply:
Then we should replace `DoBatchWithClientError` with `DoBatchWithOptions` in `CHANGELOG` too, and mention both PRs there. 
The usage of `DoBatchWithClientError` should be removed from Mimir once the vendored `dskit` is updated.